File: packages/core/src/index_single_file/scope_tree/fixtures/python/nested_scopes.py
"""
Complex Python file with nested scopes for testing
"""

# Module scope - global level
MODULE_CONSTANT = "test"

class OuterClass:
    """Class scope"""

    def __init__(self, param: str):
        """Constructor/method scope"""
        self.field = param

        # Block scope within constructor
        if param:
            local_var = "local"
            print(local_var)

    async def process_data(self, items: list[str]) -> None:
        """Method scope"""

        # Block scope - for loop
        for item in items:
            # Block scope - if statement
            if len(item) > 0:
                processed = item.upper()

                # Block scope - try-except
                try:
                    await self._handle_item(processed)
                except Exception as error:
                    print(f"Error: {error}")

                    # Block scope - nested if
                    if isinstance(error, ValueError):
                        raise RuntimeError(f"Failed to process: {error}")

    def _handle_item(self, item: str) -> None:
        """Private method scope"""

        # Nested function scope
        def helper(x: str) -> str:
            # Function scope within method
            if x.startswith("test"):
                return x.upper()
            else:
                return x.lower()

        result = helper(item)
        print(result)

    @staticmethod
    def create_default() -> 'OuterClass':
        """Static method scope"""
        return OuterClass("default")

    @classmethod
    def create_from_config(cls, config: dict) -> 'OuterClass':
        """Class method scope"""
        return cls(config.get("param", "default"))

def module_function(param: int) -> str:
    """Function scope at module level"""

    # Nested function
    def format_number(num: int) -> str:
        """Nested function scope"""

        # Block scope - match statement (Python 3.10+)
        if num == 1:
            result = "one"
            return result
        elif num == 2:
            result = "two"
            return result
        else:
            result = str(num)
            return result

    return format_number(param)

class InnerProcessor:
    """Nested class scope"""

    async def process(self, data) -> None:
        """Method scope in nested class"""

        # Block scope - while loop
        counter = 0
        while counter < 10:
            # Block scope within while
            if counter % 2 == 0:
                await self._process_even(counter)
            else:
                await self._process_odd(counter)
            counter += 1

    async def _process_even(self, num: int) -> None:
        """Method scope"""

        # Lambda function
        helper = lambda x: x * 2
        print(helper(num))

    async def _process_odd(self, num: int) -> None:
        """Method scope"""

        # Nested function with closure
        def helper(x: int) -> int:
            multiplier = 3  # Closure variable
            return x * multiplier

        print(helper(num))

# Generator function scope
def number_generator(start: int, end: int):
    """Generator function scope"""
    for i in range(start, end):
        # Block scope within generator
        if i % 2 == 0:
            yield i * 2
        else:
            yield i * 3

# Decorator function scope
def timing_decorator(func):
    """Decorator function scope"""
    import time

    def wrapper(*args, **kwargs):
        """Wrapper function scope"""
        start_time = time.time()

        try:
            result = func(*args, **kwargs)
            return result
        finally:
            end_time = time.time()
            print(f"Function took {end_time - start_time} seconds")

    return wrapper

# Context manager scope
class CustomContext:
    """Context manager class scope"""

    def __enter__(self):
        """Enter method scope"""
        print("Entering context")
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Exit method scope"""

        # Block scope - if statement
        if exc_type is not None:
            print(f"Exception occurred: {exc_val}")
            return False

        print("Exiting context normally")
        return True

# List comprehension (creates its own scope)
squared_numbers = [x**2 for x in range(10) if x % 2 == 0]

# Dictionary comprehension scope
number_map = {str(x): x**2 for x in range(5)}

# With statement creates block scope
with CustomContext() as ctx:
    # Block scope within with
    temp_var = "temporary"
    print(f"Inside context: {temp_var}")

File: packages/core/src/index_single_file/scope_tree/fixtures/rust/nested_scopes.rs
/**
 * Complex Rust file with nested scopes for testing
 */

// Module scope - global level
const MODULE_CONSTANT: &str = "test";

// Struct scope
pub struct OuterStruct {
    field: String,
}

impl OuterStruct {
    // Constructor/associated function scope
    pub fn new(param: String) -> Self {
        // Block scope within constructor
        let local_var = if !param.is_empty() {
            param.clone()
        } else {
            "default".to_string()
        };

        Self { field: local_var }
    }

    // Method scope
    pub async fn process_data(&self, items: Vec<String>) -> Result<(), Box<dyn std::error::Error>> {
        // Block scope - for loop
        for item in items {
            // Block scope - if let
            if let Some(processed) = self.process_item(&item) {
                // Block scope - match
                match self.handle_item(processed).await {
                    Ok(_) => println!("Item processed successfully"),
                    Err(e) => {
                        // Block scope within error handling
                        eprintln!("Error processing item: {}", e);

                        // Block scope - nested if
                        if e.to_string().contains("critical") {
                            return Err(e);
                        }
                    }
                }
            }
        }
        Ok(())
    }

    // Private method scope
    fn process_item(&self, item: &str) -> Option<String> {
        // Block scope - match expression
        match item.len() {
            0 => None,
            1..=10 => {
                // Block scope within match arm
                let processed = item.to_uppercase();
                Some(processed)
            },
            _ => {
                // Block scope - closure
                let transform = |s: &str| -> String {
                    // Closure scope
                    s.chars().take(10).collect()
                };
                Some(transform(item))
            }
        }
    }

    // Async method scope
    async fn handle_item(&self, item: String) -> Result<(), std::io::Error> {
        // Async block scope
        let result = async {
            // Block scope within async block
            if item.starts_with("error") {
                Err(std::io::Error::new(std::io::ErrorKind::Other, "Simulated error"))
            } else {
                Ok(())
            }
        }.await;

        result
    }
}

// Function scope at module level
fn module_function(param: i32) -> String {
    // Block scope - match
    match param {
        1 => {
            // Block scope within match arm
            let result = "one".to_string();
            result
        },
        2 => {
            let result = "two".to_string();
            result
        },
        _ => {
            // Block scope - closure
            let formatter = |x: i32| -> String {
                // Closure scope
                format!("number_{}", x)
            };
            formatter(param)
        }
    }
}

// Trait scope
pub trait DataProcessor {
    // Method signature (no scope)
    fn process(&self, data: &str) -> Result<String, String>;

    // Default implementation scope
    fn validate(&self, data: &str) -> bool {
        // Block scope - if let
        if let Ok(_) = data.parse::<i32>() {
            true
        } else {
            false
        }
    }
}

// Implementation scope
impl DataProcessor for OuterStruct {
    // Method scope
    fn process(&self, data: &str) -> Result<String, String> {
        // Block scope - if
        if data.is_empty() {
            return Err("Empty data".to_string());
        }

        // Block scope - loop
        let mut result = String::new();
        for (i, ch) in data.chars().enumerate() {
            // Block scope within loop
            if i > 0 {
                result.push('_');
            }
            result.push(ch);
        }

        Ok(result)
    }
}

// Generic function scope
fn generic_function<T, U>(input: T) -> U
where
    T: std::fmt::Display,
    U: std::str::FromStr,
    U::Err: std::fmt::Debug,
{
    // Block scope - let statement with pattern matching
    let string_repr = format!("{}", input);

    // Block scope - match
    match string_repr.parse::<U>() {
        Ok(value) => value,
        Err(e) => {
            // Block scope - panic with closure
            let error_formatter = || format!("Parse error: {:?}", e);
            panic!("{}", error_formatter());
        }
    }
}

// Enum scope
pub enum ProcessingState {
    Idle,
    Running(String),
    Completed { result: String, duration: u64 },
    Failed(String),
}

impl ProcessingState {
    // Method scope on enum
    pub fn is_active(&self) -> bool {
        // Block scope - match on self
        match self {
            ProcessingState::Idle => false,
            ProcessingState::Running(_) => true,
            ProcessingState::Completed { .. } => false,
            ProcessingState::Failed(_) => false,
        }
    }

    // Method with complex pattern matching
    pub fn get_info(&self) -> String {
        // Block scope - match with guards
        match self {
            ProcessingState::Running(task) if task.len() > 10 => {
                // Block scope with guard
                format!("Long running task: {}", &task[..10])
            },
            ProcessingState::Running(task) => format!("Running: {}", task),
            ProcessingState::Completed { result, duration } => {
                // Block scope - if expression
                let summary = if *duration > 1000 {
                    "Slow completion"
                } else {
                    "Fast completion"
                };
                format!("{}: {} (took {}ms)", summary, result, duration)
            },
            _ => "Other state".to_string(),
        }
    }
}

// Module with nested items
pub mod utils {
    // Nested module scope

    // Struct in nested module
    pub struct Helper {
        data: String,
    }

    impl Helper {
        // Method scope in nested module
        pub fn normalize(input: &str) -> String {
            // Block scope - iterator chain
            input.chars()
                .filter(|c| c.is_alphanumeric())
                .collect::<String>()
                .to_lowercase()
        }

        // Method with closure scope
        pub fn transform<F>(input: &str, transformer: F) -> String
        where
            F: Fn(&str) -> String,
        {
            // Block scope - let with complex expression
            let intermediate = input.split_whitespace()
                .map(|word| transformer(word))
                .collect::<Vec<_>>()
                .join(" ");

            intermediate
        }
    }

    // Nested function scope
    pub fn format_duration(millis: u64) -> String {
        // Block scope - match with ranges
        match millis {
            0..=999 => format!("{}ms", millis),
            1000..=59999 => {
                // Block scope - arithmetic
                let seconds = millis / 1000;
                let remaining_ms = millis % 1000;
                format!("{}.{}s", seconds, remaining_ms / 100)
            },
            _ => {
                // Block scope - complex calculation
                let minutes = millis / 60000;
                let seconds = (millis % 60000) / 1000;
                format!("{}m{}s", minutes, seconds)
            }
        }
    }
}

// Macro definition scope
macro_rules! debug_scope {
    ($expr:expr) => {
        {
            // Block scope within macro
            let result = $expr;
            println!("Debug: {} = {:?}", stringify!($expr), result);
            result
        }
    };
}

// Function using macro
fn test_macro() {
    // Block scope
    let value = debug_scope!(42 + 8);
    println!("Final value: {}", value);
}

// Tests module
#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_outer_struct() {
        // Test function scope
        let instance = OuterStruct::new("test".to_string());
        let items = vec!["item1".to_string(), "item2".to_string()];

        // Block scope - match on Result
        match instance.process_data(items).await {
            Ok(_) => println!("Test passed"),
            Err(e) => panic!("Test failed: {}", e),
        }
    }

    #[test]
    fn test_processing_state() {
        // Test function scope
        let state = ProcessingState::Running("test_task".to_string());

        // Block scope - assertion
        assert!(state.is_active());

        // Block scope - let with pattern
        let info = state.get_info();
        assert!(info.contains("Running"));
    }
}